ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -523,24 +528,104 @@ KEXT_STATIC int HandleFileOpOperation(
                 goto CleanupAndReturn;
             }
         }
-        else
+    }
+    else if (KAUTH_FILEOP_LINK == action)
+    {
+        const char* fromPath = reinterpret_cast<const char*>(arg0);
+        const char* newPath = reinterpret_cast<const char*>(arg1);
+        
+        // TODO(Mac): We need to handle failures to lookup the vnode.  If we fail to lookup the vnode
+        // it's possible that we'll miss notifications
+        errno_t toErr = vnode_lookup(newPath, 0 /* flags */, &currentVnode, context);
+        if (0 != toErr)
         {
-            PerfSample hardLinkSample(&perfTracer, PrjFSPerfCounter_FileOp_HardLinkCreated);
+            KextLog_Error("HandleFileOpOperation: vnode_lookup failed, errno %d for path '%s'", toErr, newPath);
+            goto CleanupAndReturn;
+        }
         
-            int kauthResult;
-            int kauthError;
-            if (!ProviderMessaging_TrySendRequestAndWaitForResponse(
-                    root,
-                    MessageType_KtoU_NotifyFileHardLinkCreated,
-                    currentVnode,
-                    vnodeFsidInode,
-                    newPath,
-                    pid,
-                    procname,
-                    &kauthResult,
-                    &kauthError))
+        // Don't expect named stream here as they can't be directly hardlinked or renamed, only the main fork can
+        assert(!vnode_isnamedstream(currentVnode));
+        
+        putCurrentVnode = true;
+        
+        bool isDirectory = (0 != vnode_isdir(currentVnode));
```

Comment:
Is it possible to hard-link directories?

If so, @jeschu1 we might need more reactionary work in the provider to do the right thing.

Reply:
> Is it possible to hard-link directories?

Yes, and macOS's Time Machine backup system actively uses this feature, so it's definitely something our kext needs to be able to handle gracefully.

Reply:
Filed #1152 for this.  

I think we'll want separate messages for File/Directory hardlink messages to user-mode (like we have for renames) but I think we can scope this PR to just the file hardlink message.